torch/distributed/distributed_c10d.py

```diff
@@ -73,7 +76,23 @@ def __new__(cls, name):
                              "Gloo or MPI backend for collective operations "
                              "on CPU tensors.")
         elif value == Backend.UNDEFINED:
-            raise ValueError("Invalid backend: '{}'".format(name))
+            try:
+                if name not in sys.modules:
+                    # using "pytorch_" prefix to avoid module naming conflicts/confusion 
+                    third_lib = importlib.import_module("pytorch_{}".format(name))
+                else:
+                    third_lib = sys.modules["pytorch_{}".format(name)]
+
+                pg_name = "ProcessGroup{}".format(name.upper())
```

Comment:
Upper shouldn't be needed here. We can define a getter function on the module that returns the class instead.

Reply:
@pietern May I get you clarification on this comment? do you mean we shouldn't assume the exposed instantiation interface name is ProcessGroupXYZ and should get the name from extension?  IMHO anyway we have to have assumption (or naming appointment) on the name unless there is a query interface. but the query interface's name is also an appointment...

Do I misunderstand something?

Reply:
No that's my intent indeed.

Following what you implemented now, it will first try to call `name()` on the module to query for the right name. If that doesn't work, it'll try to get the class `ProcessGroupFoo`. This seems like the proper ordering to me.